src/app/util/ember-compatibility-functions.cpp

```diff
@@ -340,5 +344,121 @@ CHIP_ERROR ReadSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVWriter * ap
+    dataLen = static_cast<uint16_t>(len + sizeof(len));
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR prepareWriteData(EmberAfAttributeType expectedType, TLV::TLVReader & aReader, uint16_t & dataLen)
+{
+    switch (BaseType(expectedType))
+    {
+    case ZCL_BOOLEAN_ATTRIBUTE_TYPE: // Boolean
+        return numericTlvDataToAttributeBuffer<bool>(aReader, dataLen);
+    case ZCL_INT8U_ATTRIBUTE_TYPE: // Unsigned 8-bit integer
+        return numericTlvDataToAttributeBuffer<uint8_t>(aReader, dataLen);
+    case ZCL_INT16U_ATTRIBUTE_TYPE: // Unsigned 16-bit integer
+        return numericTlvDataToAttributeBuffer<uint16_t>(aReader, dataLen);
+    case ZCL_INT32U_ATTRIBUTE_TYPE: // Unsigned 32-bit integer
+        return numericTlvDataToAttributeBuffer<uint32_t>(aReader, dataLen);
+    case ZCL_INT64U_ATTRIBUTE_TYPE: // Unsigned 64-bit integer
+        return numericTlvDataToAttributeBuffer<uint64_t>(aReader, dataLen);
+    case ZCL_INT8S_ATTRIBUTE_TYPE: // Signed 8-bit integer
+        return numericTlvDataToAttributeBuffer<int8_t>(aReader, dataLen);
+    case ZCL_INT16S_ATTRIBUTE_TYPE: // Signed 16-bit integer
+        return numericTlvDataToAttributeBuffer<int16_t>(aReader, dataLen);
+    case ZCL_INT32S_ATTRIBUTE_TYPE: // Signed 32-bit integer
+        return numericTlvDataToAttributeBuffer<int32_t>(aReader, dataLen);
+    case ZCL_INT64S_ATTRIBUTE_TYPE: // Signed 64-bit integer
+        return numericTlvDataToAttributeBuffer<int64_t>(aReader, dataLen);
+    case ZCL_OCTET_STRING_ATTRIBUTE_TYPE: // Octet string
+    case ZCL_CHAR_STRING_ATTRIBUTE_TYPE:  // Char string
+        return stringTlvDataToAttributeBuffer<uint8_t>(aReader, dataLen);
+    case ZCL_LONG_OCTET_STRING_ATTRIBUTE_TYPE: // Long octet string
+    case ZCL_LONG_CHAR_STRING_ATTRIBUTE_TYPE:  // Long char string
+        return stringTlvDataToAttributeBuffer<uint16_t>(aReader, dataLen);
+    default:
+        ChipLogError(DataManagement, "Attribute type %d not handled", static_cast<int>(expectedType));
+        return CHIP_ERROR_INVALID_DATA_LIST;
+    }
+}
+} // namespace
+
+CHIP_ERROR WriteSingleClusterData(ClusterInfo & aClusterInfo, TLV::TLVReader & aReader, WriteHandler * apWriteHandler)
+{
+    AttributePathParams attributePathParams;
+    attributePathParams.mNodeId     = aClusterInfo.mNodeId;
+    attributePathParams.mEndpointId = aClusterInfo.mEndpointId;
+    attributePathParams.mClusterId  = aClusterInfo.mClusterId;
+    attributePathParams.mFieldId    = aClusterInfo.mFieldId;
+    attributePathParams.mFlags.Set(AttributePathParams::Flags::kFieldIdValid);
+
+    CHIP_ERROR readError               = CHIP_NO_ERROR;
+    EmberAfAttributeType attributeType = ZCL_UNKNOWN_ATTRIBUTE_TYPE;
+
+    uint16_t dataLen     = 0;
+    EmberAfStatus status = emberAfReadAttribute(aClusterInfo.mEndpointId, aClusterInfo.mClusterId, aClusterInfo.mFieldId,
+                                                CLUSTER_MASK_SERVER, nullptr, 0, &attributeType); // write?
```

Comment:
So we so assume `CLUSTER_MASK_SERVER` even for client clusters ?

Also,  do we really need to read the attribute just to get the attribute type ?

Reply:
We should always check the value we received is the value we want.

i.e. Writing an integer value to a string value should be rejected.

Reply:
From the spec, seems we no longer have "client" clusters, only cluster clients --> endpoint does not present on clients (related spec discussion: https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/3258)

Reply:
Per docstring of emberAfReadAttribute

```
Both dataPtr and dataType may be NULL, signifying that either
 * value or type is not desired.
```

It is a desired behavior when passing a nullptr as dataptr, in this case, the data read & copy will be skipped.